nexusserver/server.go

```diff
@@ -0,0 +1,463 @@
+		CompleteOperation(context.Context, *CompletionRequest) error
+	}
+
+	// Options for creating an HTTP handler from a given [CompletionHandler].
+	CompletionOptions struct {
+		// Handler for handling completion requests.
+		Handler CompletionHandler
+		// A stuctured logging handler.
+		// Defaults to logging with text format to stderr at info level.
+		LogHandler slog.Handler
+		// Optional marshaler for marshaling objects to JSON.
+		// Defaults to output with indentation.
+		Marshaler nexusapi.Marshaler
+	}
+
+	completionHTTPHandler struct {
+		baseHTTPHandler
+		handler CompletionHandler
+	}
+)
+
+const (
+	headerOperationState = nexusapi.HeaderOperationState
+	headerContentType    = nexusapi.HeaderContentType
+	contentTypeJSON      = nexusapi.ContentTypeJSON
+)
+
+func newBadRequestError(format string, args ...any) *HandlerError {
+	return &HandlerError{
+		StatusCode: http.StatusBadRequest,
+		Failure: &nexusapi.Failure{
+			Message: fmt.Sprintf(format, args...),
+		},
+	}
+}
+
+// Error implements the error interface.
+func (e *HandlerError) Error() string {
+	return fmt.Sprintf("%d: %s", e.StatusCode, e.Failure.Message)
+}
+
+func (r *OperationResponseAsync) applyToStartResponse(writer http.ResponseWriter, handler *httpHandler) {
+	info := nexusapi.OperationInfo{
+		ID:    r.OperationID,
+		State: nexusapi.OperationStateRunning,
+	}
+	bytes, err := handler.marshaler(info)
+	if err != nil {
+		handler.logger.Error("failed to serialize operation info", "error", err)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	writer.Header().Set(headerContentType, contentTypeJSON)
+	writer.WriteHeader(http.StatusCreated)
+
+	if _, err := writer.Write(bytes); err != nil {
+		handler.logger.Error("failed to write response body", "error", err)
+	}
+}
+
+func (r *OperationResponseAsync) applyToGetResultResponse(writer http.ResponseWriter, handler *httpHandler) {
+	writer.Header().Set(headerOperationState, string(nexusapi.OperationStateRunning))
+	writer.WriteHeader(http.StatusNoContent)
+}
+
+func (r *OperationResponseSync) applyToStartResponse(writer http.ResponseWriter, handler *httpHandler) {
+	header := writer.Header()
+	for k, v := range r.Header {
+		header[k] = v
+	}
+	if closer, ok := r.Body.(io.Closer); ok {
+		defer closer.Close()
+	}
+	if _, err := io.Copy(writer, r.Body); err != nil {
+		handler.logger.Error("failed to write response body", "error", err)
+	}
+}
+
+func (r *OperationResponseSync) applyToGetResultResponse(writer http.ResponseWriter, handler *httpHandler) {
+	writer.Header().Set(headerOperationState, string(nexusapi.OperationStateSucceeded))
+	r.applyToStartResponse(writer, handler)
+}
+
+// NewJSONOperationResponseSync constructs an [OperationResponseSync], setting the proper Content-Type header.
+// Marhsals the provided value to JSON using [nexusapi.DefaultMarshaler].
+func NewJSONOperationResponseSync(v any) (*OperationResponseSync, error) {
+	b, err := nexusapi.DefaultMarshaler(v)
```

Comment:
I think operation response should just have both `BodyValue` and `BodyReader` as mutually exclusive